includes/class-wcs-action-scheduler-customer-notifications.php

```diff
@@ -0,0 +1,290 @@
+		$dt = new DateTime( $datetime, new DateTimeZone( 'UTC' ) );
+		$dt->sub( new DateInterval( "PT{$this->get_hours_offset( $subscription )}H" ) );
+
+		return $dt->getTimestamp();
+	}
+
+	public function schedule_trial_ending_notification( $subscription ) {
+		$trial_end = $subscription->get_date( 'trial_end' );
+		$timestamp = $this->sub_time_offset( $trial_end, $subscription );
+
+		$this->schedule_notification(
+			$subscription,
+			'woocommerce_scheduled_subscription_customer_notification_trial_expiration',
+			$timestamp
+		);
+	}
+
+	public function schedule_expiry_notification( $subscription ) {
+		$subscription_end = $subscription->get_date( 'end' );
+		$timestamp        = $this->sub_time_offset( $subscription_end, $subscription );
+
+		$this->schedule_notification(
+			$subscription,
+			'woocommerce_scheduled_subscription_customer_notification_expiration',
+			$timestamp
+		);
+	}
+
+	public function schedule_payment_notification( $subscription ) {
+
+		//TODO: For end of trial, should we schedule payment notification? Let's say no.
+		if ( $subscription->get_date( 'trial_end' ) ) {
+			return;
+		}
+
+		$next_payment = $subscription->get_date( 'next_payment' );
+		$timestamp    = $this->sub_time_offset( $next_payment, $subscription );
+
+		// Can manual vs automatic payment change until it runs?
+		if ( $subscription->is_manual() ) {
+			$this->schedule_notification(
+				$subscription,
+				'woocommerce_scheduled_subscription_customer_notification_manual_renewal',
+				$timestamp
+			);
+		} else {
+			$this->schedule_notification(
+				$subscription,
+				'woocommerce_scheduled_subscription_customer_notification_auto_renewal',
+				$timestamp
+			);
+		}
+	}
+
+	/**
+	 * Maybe set a schedule action if the new date is in the future
+	 *
+	 * @param object $subscription An instance of a WC_Subscription object
+	 * @param string $date_type Can be 'trial_end', 'next_payment', 'payment_retry', 'end', 'end_of_prepaid_term' or a custom date type
+	 * @param string $datetime A MySQL formatted date/time string in the GMT/UTC timezone.
+	 */
+	public function update_date( $subscription, $date_type, $datetime ) {
+
+		if ( in_array( $date_type, $this->get_date_types_to_schedule(), true ) ) {
+
+			if ( $subscription->get_date( 'trial_end' ) ) {
+				$this->schedule_trial_ending_notification( $subscription );
+			}
+
+			if ( $subscription->get_date( 'end' ) ) {
+				$this->schedule_expiry_notification( $subscription );
+			}
+
+			if ( $subscription->get_date( 'next_payment' ) ) {
+				$this->schedule_payment_notification( $subscription );
+			}
+		}
+	}
+
+	/**
+	 * Delete a date from the action scheduler queue
+	 *
+	 * @param object $subscription An instance of a WC_Subscription object
+	 * @param string $date_type Can be 'trial_end', 'next_payment', 'end', 'end_of_prepaid_term' or a custom date type
+	 */
+	public function delete_date( $subscription, $date_type ) {
+		$this->update_date( $subscription, $date_type, '0' );
```

Comment:
Just noting that the 3rd param isn't used in `update_date()`. It might still work though. Just wanted to flag as I skimmed :)

Reply:
What do you mean it might still work? 😄 
I'm old-school, so if a method is declared with 3 arguments, I provide 3 arguments. 

Jokes aside, I removed the call as I updated the logic when notifications get scheduled.

Reply:
So I've all but confirmed the issue above.

If you purchase a subscription and then delete the next payment date, the notification remains scheduled where there is no next payment date scheduled. 

ie Using the code snippet below. Deletes the next payment date but there remains a scheduled action.

```php
$subscription = wcs_get_subscription( 9276 );
$subscription->delete_date( 'next_payment' );
```
<p align="center">
<img width="304" alt="Screenshot 2024-09-20 at 4 11 59 pm" src="https://github.com/user-attachments/assets/509edaf2-1547-488a-a5be-4877212e203f">
</br>
<sup>No next payment date set in Edit subscription screen schedule metabox</sup>
</p> 

<p align="center">
<img width="1193" alt="Screenshot 2024-09-20 at 4 15 56 pm" src="https://github.com/user-attachments/assets/43207583-ff08-4d62-885f-c40a00dfdb39">
</br>
<sup>The only scheduled event for this subscription is the notification - no next payment scheduled</sup>
</p>